keps/sig-storage/20200217-namespace-transfer.md

```diff
@@ -0,0 +1,283 @@
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+This KEP proposes an API to securely transfer Kubernetes storage resources between namespaces.
+The initial implementation will focus on PersistentVolumeClaims.  But the API could be applied
+to other resource types as compelling use cases come up.
+
+`StorageTransferRequest` resources are created in the target namespace.
+`StorageTransferApproval` resources are created in the source namespace.
+When a controller detects matching request/approval resources, a transfer in initiated.
+
+When transferring a PersistentVolumeClaim, the associated PersistentVolume will be updated,
+the source PersistentVolumeClaim will be deleted, and the target PersistentVolumeClaim will be created.
+The transfer simply deals with Kubernetes API resources.
+No data on the physical PersistentVolume is accessed/copied. 
+
+## Motivation
+
+Give Kubernetes users the flexibility to easily and securely share data between namespaces.
+
+### Goals
+
+- Define an API for transferring persistent storage resources between namespaces.
+- The API should be compatible with as many existing storage provisioners as possible.
+
+### Non-Goals
+
+- Transfer of resources other than PersistentVolumeClaims will not be discussed.
+- Implementation details only discussed in the context of influencing the API definition.
+
+## Proposal
+
+`StorageTransferRequest` resource:
+
+```yaml
+apiVersion: v1alpha1
+kind: StorageTransferRequest
+metadata:
+  name: transfer-foo
+  namespace: target-namespace
+spec:
+  source:
+    name: foo
+    namespace: source-namespace
+    kind: PersistentVolumeClaim
+  targetName: bar #this is optional
+```
+
```

Comment:
I don't know that it's necessary, but do we want to consider requiring a secret associated with this?  It may be overkill because it's an active operation on both sides and it will only transfer to the destination specified by the source; but just something to think about (briefly).  Maybe that's an implementation detail and we address it later.